File: 1406-stone-game-iii/1406-stone-game-iii.cpp
class Solution {
public:
    string stoneGameIII(vector<int>& stoneValue) {
        int n = stoneValue.size();

        vector<int>dp(n+1 , 0);

        for(int i = n-1 ; i >=0 ; i--){
            dp[i] = INT_MIN;
            int take_sum = 0;

            for(int k = 1 ; k < 4 ; k++){
                if(i+k <= n){
                    take_sum += stoneValue[i+k-1];
                    dp[i] = max(dp[i] , take_sum - dp[i+k]);
                }
            }
        }

        if(dp[0] >0){
            return "Alice";
        }
        else if(dp[0] < 0){
            return "Bob";
        }
        else{
            return "Tie";
        }
    }
};

File: 1406-stone-game-iii/1406-stone-game-iii.py
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        
        dp = [0]*(n + 1)

        for i in range(n-1 , -1 , -1):
            dp[i] = float('-inf')
            take_sum = 0

            for k in range(1 , 4):
                if i + k <= n:
                    take_sum += stoneValue[i + k - 1]
                    dp[i] = max(dp[i] , take_sum - dp[i+ k])
            

        if dp[0] >0:
            return "Alice"
        elif dp[0] < 0:
            return "Bob"
        else:
            return 'Tie'